File: 1249MinimumRemoveToMakeValidParentheses/LC1249.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        o =[]
        c = set()

        for i,v in enumerate(s):
            if v not in '()':
                continue
            if v == '(':
                o.append(i)
            elif not o:# ')' appear before '('
                c.add(i)
            else: # v ==')'
                o.pop()
        c = c.union(set(o))
        strs =''
        for i, v in enumerate(s):
            if i not in c:
                strs += v
        return strs
                
# TC:O(N)
# SC:O(N) o and c

File: 1249MinimumRemoveToMakeValidParentheses/LC1249.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        Stack<Integer> open = new Stack<Integer>();
        Set<Integer> close = new HashSet<>();
        for (int i = 0; i< s.length(); i++) {
            if (s.charAt(i) == '(') {
                open.push(i);
            } else if (s.charAt(i) == ')') {
                if (open.isEmpty()) {
                    close.add(i);
                }else {
                    open.pop();}
            } 
        }
        while (!open.isEmpty()) close.add(open.pop());
        
        StringBuilder str = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            if (!close.contains(i)) {
                str.append(s.charAt(i));
            }
        }
        return str.toString();
        
    }
}